### ISSUE TITLE
identifying disconnected users 

### ISSUE THREAD

**Issue by Hretic** [auth0-community/auth0-socketio-jwt#170]
i have a php/laravel cms and i have a simple nodejs game server which is basically a loop generating some numbers 

i connect my php backend to nodejs via socketio and use `socketio-jwt` to identify the user 

my client side (php/laravel)

php 

` $userToken = JWTAuth::customClaims(['userid' => Auth::user()->id, 'name' => Auth::user()->name, 'avatar' => asset_url('image/avatar-default.png')])->fromUser(Auth::user() );`

html/js

```
        var socket = io.connect("http://localhost:666");

        socket.on('connect', function () {
            socket.emit('authenticate', {token: '{{$userToken}}'});
        });
        socket.on('authenticated', function () {
            console.log('Authenticated');
        });
        socket.on('unauthorized', function (data) {
            console.log('Unauthorized, error msg: ' + data.message);
        });

```

my serverside 

```
const _online_users = { };
io.sockets

    .on('connection', socketioJwt.authorize({
        secret:  JWT_SECRET,
        timeout: 15000
    }))

    .on('authenticated', function(socket) {

        _online_users[socket.decoded_token.userid] = {
            name : socket.decoded_token.name ,
            avatar : socket.decoded_token.avatar ,
        }

        io.sockets.emit('update_online_users' ,  _online_users  );
            socket.on('disconnect', function() {
                 console.log(`----- @@disconnect -----`)
            });
    })  ;

```

as you can see i have a object called `_online_users ` and i store authenticated users in this object and then i send it to the clients so they know who is online 
 `   io.sockets.emit('update_online_users' ,  _online_users  );`

here is the problem , when user get disconnected 

```
            socket.on('disconnect', function() {
                 console.log(`----- @@disconnect -----`)
            });
```

i have to update my `_online_users `  object and remove disconnected user .... how should i go about this ? i was thinking maybe i can store the token itself in the `_online_users `  

```
        _online_users[socket.decoded_token.userid] = {
           token  : socket.token ,
            name : socket.decoded_token.name ,
            avatar : socket.decoded_token.avatar ,
        }
```

and when user get disconnected i get the disconnected token from socket and remove the user from object by that  token 

of course this is all theory ! im not sure if thats the way to go .... first of all i cant access the token it self from socket ! 

or lets say one of the users sends another request to node server , how can i identify the user sending the request ? 

```
    .on('authenticated', function(socket) {

            socket.on('somaction', function() {
                 console.log(` who is this guy ? `)
            });

    })

```

is there anything unique in`socket.decoded_token` that i can use as id ?  if so i can store it in the online users send it back and forth when user  requests something 

basically im lost and appreciate any pointers (im new to node in case u didnt notice ! )

**Comment by Hretic** [comment]
ok it was simple i should have seen this !



```
        socket.on('disconnect', function() {
            console.log(`----- @@disconnect -----`);

            delete _online_users[socket.decoded_token.userid];

            io.sockets.emit('update_online_users', _online_users);
        });

```
